d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,709 @@
+
+        for (FilterChain filterChain : listener.getFilterChainsList())
+        {
+          for (Filter filter : filterChain.getFiltersList())
+          {
+            if (filter.hasTypedConfig() && filter.getTypedConfig()
+                .getTypeUrl()
+                .equals(TYPE_URL_HTTP_CONNECTION_MANAGER))
+            {
+              HttpConnectionManager httpConnectionManager = filter.getTypedConfig().unpack(HttpConnectionManager.class);
+              ldsUpdates.put(listener.getName(), new LdsUpdate(httpConnectionManager));
+            }
+          }
+        }
+      } catch (InvalidProtocolBufferException e)
+      {
+        _log.warn("Failed to unpack LDS response", e);
+        errors.add("Failed to unpack LDS response");
+      }
+    }
+
+    if (errors.isEmpty())
+    {
+      _adsStream.sendAckRequest(responseData.getResourceType(), responseData.getNonce());
+    } else
+    {
+      String errorDetail = Joiner.on('\n').join(errors);
+      _adsStream.sendNackRequest(responseData.getResourceType(), responseData.getNonce(), errorDetail);
+    }
+
+    for (Map.Entry<String, LdsUpdate> entry : ldsUpdates.entrySet())
+    {
+      String resourceName = entry.getKey();
+      LdsUpdate ldsUpdate = entry.getValue();
+      ResourceSubscriber subscriber = _ldsResourceSubscribers.get(resourceName);
+      if (subscriber != null)
+      {
+        subscriber.onData(ldsUpdate);
+      }
+    }
+  }
+
+  private void handleRdsResponse(DiscoveryResponseData responseData)
+  {
+    Map<String, RdsUpdate> rdsUpdates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : responseData.getResourcesList())
+    {
+      try
+      {
+        RouteConfiguration routeConfig = resource.getResource().unpack(RouteConfiguration.class);
+        List<VirtualHost> virtualHostList = routeConfig.getVirtualHostsList();
+        Map<String, Value> serviceProperties = null;
+        for (VirtualHost virtualHost : virtualHostList)
+        {
+          for (Route route : virtualHost.getRoutesList())
+          {
+            Map<String, Struct> metadataMap = route.getMetadata().getFilterMetadataMap();
+            if (metadataMap.containsKey(D2_SERVICE_PROPERTIES_KEY))
+            {
+              serviceProperties = metadataMap.get(D2_SERVICE_PROPERTIES_KEY).getFieldsMap();
+            }
+          }
+        }
+        if (serviceProperties == null)
+        {
+          _log.warn("Missing service properties in RDS response");
+          errors.add("Missing service properties in RDS response");
+          continue;
+        }
+        rdsUpdates.put(routeConfig.getName(), new RdsUpdate(routeConfig.getVirtualHostsList(), serviceProperties));
+      } catch (InvalidProtocolBufferException e)
+      {
+        _log.warn("Failed to unpack RDS response", e);
+        errors.add("Failed to unpack RDS response");
+      }
+    }
+
+    if (errors.isEmpty())
```

Comment:
The ACK / NAK part and the watcher invocation part are all similar in the handle*dsUpdate funtions, can we consolidate them?

Reply:
I think that will be one use case of your ResourceUpdate interface :)

Reply:
Good point. I refactored the handle*dsUpdate to make it more concise